tensorflow/core/util/sorter.h

```diff
@@ -0,0 +1,130 @@
+/* Copyright 2015 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#ifndef TENSORFLOW_UTIL_SORTER_H_
+#define TENSORFLOW_UTIL_SORTER_H_
+
+#include <algorithm>
+#include <functional>
+#include <vector>
+
+#include "tensorflow/core/lib/core/blocking_counter.h"
+#include "tensorflow/core/lib/core/threadpool.h"
+#include "tensorflow/core/platform/logging.h"
+
+namespace tensorflow {
+
+namespace {
+
+int divup(int a, int b) { return (a + (b - 1)) / b; }
+
+// Recursively re-order x, which makes all elements within [first, pos) less
+// than those in [pos, last).
+template <typename T, class Compare>
+void DivideAtPos(std::vector<T>& x, Compare& comp, int first, int last,
```

Comment:
`DivideAtPost` takes `O(n)` (when `n = last-first`) and potentially more. This means this parallel QSort will still take `O(n log n)` on the critical path. How much speedup does this get over serial QSort?